### PR TITLE
Interchain Tests

### PR DESCRIPTION
**Pull request by anhductn2001**
Interchaintest orchestrates Go tests that utilize Docker containers for multiple IBC-compatible blockchains.
It allows users to quickly spin up custom testnets and dev environments to test IBC, chain infrastructures, smart contracts, etc.
1. We will build a Migaloo image and use it for testing.
2. Setup interchain (ChainConfig, OverrideConfig, numberOfVal, Name of Chain, ...)
3. Write Testcase (StartChain, PushWasmCode, IBC Transfers, ...)

### CODE REVIEW

**Comment by faddat** [comment]
Thanks for taking this issue! Please let me know if you need any help

**Comment by faddat** [comment]
yes I can, thanks for asking @anhductn2001

**Comment by faddat** [comment]
@anhductn2001 sir here's your image:

https://github.com/White-Whale-Defi-Platform/migaloo-chain/pkgs/container/migaloo-chain-ictest

use the ghcr path you can find there :)

**Comment by anhductn2001** [comment]
Please public it @faddat. The status is private now. TY!

**Comment by faddat** [comment]
ahhh gotcha.  1 sec

**Comment by faddat** [comment]
I think you need to ask @Sen-Com in slack sir

**Comment by anhductn2001** [comment]
Yea, I will

File: .gitignore [hunk 1]
```diff
@@ -3,4 +3,5 @@ bin
 .ash_history
 
 
-
+go.work
```
Comment (faddat) [inline .gitignore:6]: we will commit go.work but not go.sum

File: go.mod [hunk 2]
```diff
@@ -194,4 +194,5 @@ replace (
 	// Downgraded to avoid bugs in following commits which caused simulations to fail.
 	github.com/syndtr/goleveldb => github.com/syndtr/goleveldb v1.0.1-0.20210819022825-2ae1ddf74ef7
 
+	github.com/terra-money/alliance v0.2.0 => github.com/notional-labs/alliance v1.0.1-0.20230525040444-6563ecc53ade
```
[the remark this answers is not in the document]
Comment (anhductn2001) [reply go.mod]: I realized that alliance forgot to `go mod tidy`, so lib `github.com/coinbase/rosetta-sdk-go/types/` has not updated. And I only tidy for this here: https://github.com/notional-labs/alliance/commit/6563ecc53ade4de8c288195c55cf7edbaf046ae1

**Comment by anhductn2001** [comment]
> hey I think I am going to pull the docker image out of here, then merge it to main...

Hmm, I am thinking about that. There seems to be no response from someone who has the permission to manage ghcr package yet

**Comment by faddat** [comment]
it's because that person is me 🤦

File: .github/workflows/interchaintest.yml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+          context: .
+          push: true
+          platforms: linux/amd64,linux/arm64
+          tags: |
+            ghcr.io/white-whale-defi-platform/migaloo-chain-ictest:latest
+  test-start-cosmos-chain:
+    runs-on: ubuntu-latest
+    needs: build-and-push-image
+    steps:
+      - name: Set up Go 1.19
```
Comment (faddat) [inline .github/workflows/interchaintest.yml:43]: 1.20

Comment (anhductn2001) [reply .github/workflows/interchaintest.yml:43]: Ok, sir

File: go.mod (hunk 2 above)
[the remark this answers is not in the document]
Comment (faddat) [reply go.mod]: ahhhhhhhhhh okkkkkkk I see I see I see that is where the problem comes in!

[the remark this answers is not in the document]
Comment (faddat) [reply go.mod]: lets see if we can use the main branch of alliance successfully

[the remark this answers is not in the document]
Comment (faddat) [reply go.mod]: If not, please make a PR to the terra-money repo with your fix -- the rosetta thing was for sure super annoying.  Thanks!

[the remark this answers is not in the document]
Comment (anhductn2001) [reply go.mod]: This PR here: https://github.com/terra-money/alliance/pull/202 . I dont know why they dont merge yet!

**Comment by anhductn2001** [comment]
We need it pass: https://github.com/White-Whale-Defi-Platform/migaloo-chain/issues/179 @faddat

**Comment by faddat** [comment]
@anhductn2001 this is going to likely fall outside my best places, can you please check with @catShaark and @vuong177 ?

Thanks :)

**Comment by faddat** [comment]
lets do the go workspace.  I like them

**Comment by anhductn2001** [comment]
Workspace should be add manual in local env, sir. We dont need add go.work on repo, it should a go.work.example

**Comment by faddat** [comment]
do like this:

https://github.com/osmosis-labs/osmosis/blob/main/go.work

**Comment by anhductn2001** [comment]
Finally, I have done with #179 @faddat

**Comment by anhductn2001** [comment]
> do like this:
> 
> https://github.com/osmosis-labs/osmosis/blob/main/go.work
> 
> commit go.work but don't commit go.work.sum

We should temporarily remove it @faddat , because it causes errors when building binary. This is related to the dep version in interchaintest

**Comment by anhductn2001** [comment]
Can you review this @faddat?

**Comment by faddat** [comment]
Yep, please call me next time you need a review if it doesn't happen promptly :)

**Review by faddat — APPROVED** [review]
looks like a basic implementation we will want to add upgrading testing, but -- great start!

**Pull request merged**